src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -78,46 +62,57 @@ public byte id() {
         }
     }
 
-    public interface Custom {
+    public static final String TYPE = "cluster";
 
-        interface Factory<T extends Custom> {
+    public static final Factory FACTORY = new Factory();
 
-            String type();
+    public static class Factory extends AbstractCompositeFactory<ClusterState> {
 
-            T readFrom(StreamInput in) throws IOException;
-
-            void writeTo(T customState, StreamOutput out) throws IOException;
+        @Override
+        public ClusterState fromParts(long version, String uuid, ImmutableOpenMap.Builder<String, ClusterStatePart> parts) {
+            return new ClusterState(version, uuid, parts.build());
+        }
 
-            void toXContent(T customState, XContentBuilder builder, ToXContent.Params params);
+        @Override
+        public String partType() {
+            return TYPE;
         }
     }
 
-    private final static Map<String, Custom.Factory> customFactories = new HashMap<>();
-
-    /**
-     * Register a custom index meta data factory. Make sure to call it from a static block.
-     */
-    public static void registerFactory(String type, Custom.Factory factory) {
-        customFactories.put(type, factory);
+    static {
+        FACTORY.registerFactory(ClusterName.FACTORY);
```

Comment:
I wonder if we could use an SPI loader instead of these factories just like lucene does this for its codecs?